Possible back end issues with new unified login?

##### Brief description of problem/feature
I know that @yeisenberg was trying to look at the Mendota admin page earlier today. He said that he got it to load once, surprisingly! But that it failed later. It has always failed for me since unified login, timing out at 60 seconds before it finishes loading.

But then I noticed that the Mendota server restarted itself this morning as well. My guess it was while Yochai or I were trying to access the admin page. Here are some of the errors that I see in the logs (pasted only the part of the stack trace that includes references to our code):
```
2024-11-04 11:18:24,426 - [ERROR] - from application in play-akka.actor.default-dispatcher-2480
Internal server error, for (HEAD) [/signIn] ->

play.api.Application$$anon$1: Execution exception[[SQLException: Timed out waiting for a free available connection.]]
...
        at scala.slick.jdbc.PlayDatabase.withTransaction(PlayDatabase.scala:6) ~[com.typesafe.play.play-slick_2.10-0.8.1.jar:0.8.1]
        at models.daos.slick.DBTableDefinitions$UserTable$.find(DBTableDefinitions.scala:66) ~[sidewalk-webpage.sidewalk-webpage-8.0.0.jar:8.0.0]
        at controllers.UserController$$anonfun$2.apply(UserController.scala:117) ~[sidewalk-webpage.sidewalk-webpage-8.0.0.jar:8.0.0]
        at controllers.UserController$$anonfun$2.apply(UserController.scala:117) ~[sidewalk-webpage.sidewalk-webpage-8.0.0.jar:8.0.0]
        at scala.Option.getOrElse(Option.scala:120) ~[org.scala-lang.scala-library-2.10.7.jar:na]
        at controllers.UserController.logPageVisit(UserController.scala:117) ~[sidewalk-webpage.sidewalk-webpage-8.0.0.jar:8.0.0]
        at controllers.UserController$$anonfun$signIn$1.apply(UserController.scala:35) ~[sidewalk-webpage.sidewalk-webpage-8.0.0.jar:8.0.0]
        at controllers.UserController$$anonfun$signIn$1.apply(UserController.scala:33) ~[sidewalk-webpage.sidewalk-webpage-8.0.0.jar:8.0.0]
...
```

And then this:
```
2024-11-04 11:23:43,151 - [ERROR] - from play.nettyException in New I/O worker #46
Exception caught in Netty
java.lang.OutOfMemoryError: GC overhead limit exceeded
...
        at scala.slick.jdbc.PlayDatabase.withTransaction(PlayDatabase.scala:6) ~[com.typesafe.play.play-slick_2.10-0.8.1.jar:0.8.1]
        at models.daos.slick.DBTableDefinitions$UserTable$.find(DBTableDefinitions.scala:66) ~[sidewalk-webpage.sidewalk-webpage-8.0.0.jar:8.0.0]
        at controllers.UserController$$anonfun$2.apply(UserController.scala:117) ~[sidewalk-webpage.sidewalk-webpage-8.0.0.jar:8.0.0]
        at controllers.UserController$$anonfun$2.apply(UserController.scala:117) ~[sidewalk-webpage.sidewalk-webpage-8.0.0.jar:8.0.0]
        at scala.Option.getOrElse(Option.scala:120) ~[org.scala-lang.scala-library-2.10.7.jar:na]
        at controllers.UserController.logPageVisit(UserController.scala:117) ~[sidewalk-webpage.sidewalk-webpage-8.0.0.jar:8.0.0]
...
```

##### Potential solution(s)
Since the authentication info is now centralized, I wonder if we're having problems with concurrent reads/writes. One thing we might want to do is to check for any places where we're using `withTransaction` instead of `withSession` where we don't _need_ to be in the authentication code. I assume that this essentially takes out a lock on a table in order to make updates to it. But there's rarely a case where we need to use `withTransaction` if we're just doing a read. And I specifically see `scala.slick.jdbc.PlayDatabase.withTransaction` mentioned in the code above.

I could also spend some time looking into what various database connections are doing during sign in and when loading the admin page. More investigation needed!

Is this only affecting the admin page loading or will normal users be
affected too?

Sent from phone


On Mon, Nov 4, 2024 at 6:45 PM Michael Saugstad ***@***.***>
wrote:

> Brief description of problem/feature
>
> I know that @yeisenberg <https://github.com/yeisenberg> was trying to
> look at the Mendota admin page earlier today. He said that he got it to
> load once, surprisingly! But that it failed later. It has always failed for
> me since unified login, timing out at 60 seconds before it finishes loading.
>
> But then I noticed that the Mendota server restarted itself this morning
> as well. My guess it was while Yochai or I were trying to access the admin
> page. Here are some of the errors that I see in the logs (pasted only the
> part of the stack trace that includes references to our code):
>
> 2024-11-04 11:18:24,426 - [ERROR] - from application in play-akka.actor.default-dispatcher-2480
> Internal server error, for (HEAD) [/signIn] ->
>
> play.api.Application$$anon$1: Execution exception[[SQLException: Timed out waiting for a free available connection.]]
> ...
>         at scala.slick.jdbc.PlayDatabase.withTransaction(PlayDatabase.scala:6) ~[com.typesafe.play.play-slick_2.10-0.8.1.jar:0.8.1]
>         at models.daos.slick.DBTableDefinitions$UserTable$.find(DBTableDefinitions.scala:66) ~[sidewalk-webpage.sidewalk-webpage-8.0.0.jar:8.0.0]
>         at controllers.UserController$$anonfun$2.apply(UserController.scala:117) ~[sidewalk-webpage.sidewalk-webpage-8.0.0.jar:8.0.0]
>         at controllers.UserController$$anonfun$2.apply(UserController.scala:117) ~[sidewalk-webpage.sidewalk-webpage-8.0.0.jar:8.0.0]
>         at scala.Option.getOrElse(Option.scala:120) ~[org.scala-lang.scala-library-2.10.7.jar:na]
>         at controllers.UserController.logPageVisit(UserController.scala:117) ~[sidewalk-webpage.sidewalk-webpage-8.0.0.jar:8.0.0]
>         at controllers.UserController$$anonfun$signIn$1.apply(UserController.scala:35) ~[sidewalk-webpage.sidewalk-webpage-8.0.0.jar:8.0.0]
>         at controllers.UserController$$anonfun$signIn$1.apply(UserController.scala:33) ~[sidewalk-webpage.sidewalk-webpage-8.0.0.jar:8.0.0]
> ...
>
> And then this:
>
> 2024-11-04 11:23:43,151 - [ERROR] - from play.nettyException in New I/O worker #46
> Exception caught in Netty
> java.lang.OutOfMemoryError: GC overhead limit exceeded
> ...
>         at scala.slick.jdbc.PlayDatabase.withTransaction(PlayDatabase.scala:6) ~[com.typesafe.play.play-slick_2.10-0.8.1.jar:0.8.1]
>         at models.daos.slick.DBTableDefinitions$UserTable$.find(DBTableDefinitions.scala:66) ~[sidewalk-webpage.sidewalk-webpage-8.0.0.jar:8.0.0]
>         at controllers.UserController$$anonfun$2.apply(UserController.scala:117) ~[sidewalk-webpage.sidewalk-webpage-8.0.0.jar:8.0.0]
>         at controllers.UserController$$anonfun$2.apply(UserController.scala:117) ~[sidewalk-webpage.sidewalk-webpage-8.0.0.jar:8.0.0]
>         at scala.Option.getOrElse(Option.scala:120) ~[org.scala-lang.scala-library-2.10.7.jar:na]
>         at controllers.UserController.logPageVisit(UserController.scala:117) ~[sidewalk-webpage.sidewalk-webpage-8.0.0.jar:8.0.0]
> ...
>
> Potential solution(s)
>
> Since the authentication info is now centralized, I wonder if we're having
> problems with concurrent reads/writes. One thing we might want to do is to
> check for any places where we're using withTransaction instead of
> withSession where we don't *need* to be in the authentication code. I
> assume that this essentially takes out a lock on a table in order to make
> updates to it. But there's rarely a case where we need to use
> withTransaction if we're just doing a read. And I specifically see
> scala.slick.jdbc.PlayDatabase.withTransaction mentioned in the code above.
>
> I could also spend some time looking into what various database
> connections are doing during sign in and when loading the admin page. More
> investigation needed!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ProjectSidewalk/SidewalkWebpage/issues/3726>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAML55JRM5ZDDXLDEGVV2RLZ7AWLPAVCNFSM6AAAAABRFRP4VCVHI2DSMVQWIX3LMV43ASLTON2WKOZSGYZTIMRXHAZDMNI>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

--- a/app/models/daos/slick/UserDAOSlick.scala
+++ b/app/models/daos/slick/UserDAOSlick.scala
@@ -10,7 +10,7 @@ import models.label.LabelValidationTable
 import models.label.LabelTable
 import models.mission.MissionTable
 import models.user.UserTeamTable.userTeams
-import models.user.{RoleTable, TeamTable, User, UserRoleTable, UserStatTable, WebpageActivityTable}
+import models.user.{Role, RoleTable, TeamTable, User, UserRoleTable, UserStatTable, WebpageActivityTable}
 import play.api.db.slick._
 import play.api.db.slick.Config.driver.simple._
 import play.api.Play.current
@@ -150,17 +150,16 @@ object UserDAOSlick {
   val auditTaskTable = TableQuery[AuditTaskTable]
 
   /**
-   * Get all users, excluding anonymous users who haven't placed any labels or have done validations 
-   * (so admin user table isn't too big).
+   * Get all users, excluding anon users who haven't placed any labels or done any validations (to limit table size).
    */
-  def usersMinusAnonUsersWithNoLabelsAndNoValidations: Query[UserTable, DBUser, Seq] = { 
+  def usersMinusAnonUsersWithNoLabelsAndNoValidations: Query[(UserTable, RoleTable), (DBUser, Role), Seq] = {
     val anonUsersWithLabels = (for {
       _user <- userTable
       _userRole <- userRoleTable if _user.userId === _userRole.userId
       _role <- roleTable if _userRole.roleId === _role.roleId
       _label <- LabelTable.labelsWithTutorialAndExcludedUsers if _user.userId === _label.userId
       if _role.role === "Anonymous"
-    } yield _user).groupBy(x => x).map(_._1)
+    } yield (_user, _role)).groupBy(x => x).map(_._1)
 
     val anonUsersWithValidations = (for {
       _user <- userTable
@@ -168,14 +167,14 @@ object UserDAOSlick {
       _role <- roleTable if _userRole.roleId === _role.roleId
       _labelValidation <- LabelValidationTable.validationLabels if _user.userId === _labelValidation.userId
       if _role.role === "Anonymous"
-    } yield _user).groupBy(x => x).map(_._1)
+    } yield (_user, _role)).groupBy(x => x).map(_._1)
 
     val otherUsers = for {
       _user <- userTable
       _userRole <- userRoleTable if _user.userId === _userRole.userId
       _role <- roleTable if _userRole.roleId === _role.roleId
       if _role.role =!= "Anonymous"
-    } yield _user
+    } yield (_user, _role)
 
     anonUsersWithLabels.union(anonUsersWithValidations) ++ otherUsers
   }
@@ -475,10 +474,6 @@ object UserDAOSlick {
     // queries. We are using Scala Map objects instead of Slick b/c Slick doesn't create very efficient queries for this
     // use-case (at least in the old version of Slick that we are using right now).
 
-    // Map(user_id: String -> role: String).
-    val roles =
-      userRoleTable.innerJoin(roleTable).on(_.roleId === _.roleId).map(x => (x._1.userId, x._2.role)).list.toMap
-
     // Map(user_id: String -> team: String).
     val teams =
       userTeams.innerJoin(TeamTable.teams).on(_.teamId === _.teamId).map(x => (x._1.userId, x._2.name)).list.toMap
@@ -512,12 +507,12 @@ object UserDAOSlick {
       UserStatTable.userStats.map { x => (x.userId, x.highQuality) }.list.toMap
 
     // Now left join them all together and put into UserStatsForAdminPage objects.
-    usersMinusAnonUsersWithNoLabelsAndNoValidations.list.map { u =>
-      val ownValidatedCounts = validatedCounts.getOrElse(u.userId, ("", 0, 0))
+    usersMinusAnonUsersWithNoLabelsAndNoValidations.list.map { case (user, role) =>
+      val ownValidatedCounts = validatedCounts.getOrElse(user.userId, ("", 0, 0))
       val ownValidatedTotal = ownValidatedCounts._2
       val ownValidatedAgreed = ownValidatedCounts._3
 
-      val otherValidatedCounts = othersValidatedCounts.getOrElse(u.userId, (0, 0))
+      val otherValidatedCounts = othersValidatedCounts.getOrElse(user.userId, (0, 0))
       val otherValidatedTotal = otherValidatedCounts._1
       val otherValidatedAgreed = otherValidatedCounts._2
 
@@ -530,17 +525,17 @@ object UserDAOSlick {
         else otherValidatedAgreed * 1.0 / otherValidatedTotal
 
       UserStatsForAdminPage(
-        u.userId, u.username, u.email,
-        roles.getOrElse(u.userId, ""),
-        teams.get(u.userId),
-        signUpTimes.get(u.userId).flatten,
-        signInTimesAndCounts.get(u.userId).flatMap(_._1), signInTimesAndCounts.get(u.userId).map(_._2).getOrElse(0),
-        labelCounts.getOrElse(u.userId, 0),
+        user.userId, user.username, user.email,
+        role.role,
+        teams.get(user.userId),
+        signUpTimes.get(user.userId).flatten,
+        signInTimesAndCounts.get(user.userId).flatMap(_._1), signInTimesAndCounts.get(user.userId).map(_._2).getOrElse(0),
+        labelCounts.getOrElse(user.userId, 0),
         ownValidatedTotal,
         ownValidatedAgreedPct,
         otherValidatedTotal,
         otherValidatedAgreedPct,
-        userHighQuality.getOrElse(u.userId, true)
+        userHighQuality.getOrElse(user.userId, true)
       )
     }
   }

